Support ENABLE_WEB_CODECS in CMake builds

donny-dont opened the pull request:
#### 0ae33bb1d58c2241039204f03e8cbdb84d65b09a
<pre>
Support ENABLE_WEB_CODECS in CMake builds
<a href="https://bugs.webkit.org/show_bug.cgi?id=247777">https://bugs.webkit.org/show_bug.cgi?id=247777</a>

Reviewed by Philippe Normand and Youenn Fablet.

Add `ENABLE_WEB_CODECS` to the list of options for CMake ports. Append
it to the list of options in FeatureList.pm. Modified the files to do
a successful non-unified compilation on WinCairo.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.cpp:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.cpp:
* Source/WebCore/platform/VideoDecoder.cpp:
* Source/WebCore/platform/VideoEncoder.cpp:
* Source/cmake/WebKitFeatures.cmake:
* Tools/Scripts/webkitperl/FeatureList.pm:

Canonical link: <a href="https://commits.webkit.org/256694@main">https://commits.webkit.org/256694@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/480ab621facd47ff79af220f2b6cd1ad96b847dd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96504 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5758 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29577 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106037 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166391 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100486 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5942 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34501 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88878 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/102761 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102177 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/4435 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83106 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31399 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/86277 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88161 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/74304 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/87481 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/40233 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19652 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/82886 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37901 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21044 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28311 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4646 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/4168 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/43605 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/85566 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/982 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40315 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19284 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.cpp:
```diff
@@ -151,10 +151,14 @@ ExceptionOr<Ref<WebCodecsVideoFrame>> WebCodecsVideoFrame::create(CanvasImageSou
         if (!canvas->width() || !canvas->height())
             return Exception { InvalidStateError,  "Input canvas has a bad size"_s };
 
+#if ENABLE(MEDIA_STREAM)
```
Is this the right guard?

youennf replied:
We should probably change HTMLCanvasElement::toVideoFrame to be enabled either if MEDIA_STREAM or WEB_CODECS are defined.

donny-dont replied:
I can add that to the patch

donny-dont replied:
https://github.com/WebKit/WebKit/pull/6401

philn reviewed (commented):
Might be a bit early, no? Unless you have a platform backend for Sony port?
I started one for WPE and actually had a similar patch in my branch, but was waiting my backend to be in good shape before enabling this in CMake ports.

philn commented on Source/WebCore/Sources.txt:
```diff
@@ -2279,6 +2279,8 @@ platform/graphics/transforms/TransformOperations.cpp
 platform/graphics/transforms/TransformationMatrix.cpp
 platform/graphics/transforms/TranslateTransformOperation.cpp
+platform/libwebrtc/LibWebRTCVPXVideoDecoder.cpp
+platform/libwebrtc/LibWebRTCVPXVideoEncoder.cpp
```
That should be in some Platform cmake file, not the generic sources list imho.

donny-dont replied:
I agree with you but there's already a bunch of libwebrtc files in Sources.txt

donny-dont replied:
It probably makes more sense to have a SourcesLibWebRTC.txt but I'm not sure how all of that hooks into the Apple XCode build.

donny-dont replied:
@philn attempting to do it in https://github.com/WebKit/WebKit/pull/6413

donny-dont committed:
Support ENABLE_WEB_CODECS in CMake builds
https://bugs.webkit.org/show_bug.cgi?id=247777

Reviewed by Philippe Normand and Youenn Fablet.

Add `ENABLE_WEB_CODECS` to the list of options for CMake ports. Append
it to the list of options in FeatureList.pm. Modified the files to do
a successful non-unified compilation on WinCairo.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.cpp:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.cpp:
* Source/WebCore/platform/VideoDecoder.cpp:
* Source/WebCore/platform/VideoEncoder.cpp:
* Source/cmake/WebKitFeatures.cmake:
* Tools/Scripts/webkitperl/FeatureList.pm:

Canonical link: https://commits.webkit.org/256694@main
